### PR TITLE
Update workspace.py

### PR DESCRIPTION
**Pull request by jacobyoby**
background:
this pull request adds a new function safe_path_join() to help ensure that any file input or output is within a specific workspace directory. The function takes a base path and one or more additional path components, joins them together, and checks that the resulting path is within the workspace directory. If the path is outside the workspace, the function raises a ValueError.

changes:

add safe_path_join() function to ensure paths are within the workspace directory.
modify os.makedirs() call to use exist_ok=True argument to avoid FileExistsError when creating the workspace directory.
documentation:
the safe_path_join() function is documented with a docstring that explains its purpose and usage. the pull request description also includes an explanation of the function and its intended use.

test plan:
the following tests have been added to ensure that the safe_path_join() and path_in_workspace() functions work correctly:

test path_in_workspace() with a relative path and an absolute path to ensure that the function returns the expected absolute path within the workspace directory.
test safe_path_join() with a valid path to ensure that the function joins the path components correctly and returns the expected absolute path within the workspace directory.
test safe_path_join() with an invalid path to ensure that the function raises a ValueError with the expected error message.
pr quality checklist:

- [ x ] My pull request is atomic and focuses on a single change.
- [ x ] I have thoroughly tested my changes with multiple different prompts.
- [ x ] I have considered potential risks and mitigations for my changes.
- [ x ] I have documented my changes clearly and comprehensively.
- [ x ] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Comment by Pwuts** [comment]
Still working on it, or ready for review?

**Comment by jacobyoby** [comment]
i need to test.

File: tests/test_workspace.py
Comment (Pwuts) [inline tests/test_workspace.py:1]: Please move to `tests/unit`

File: tests/test_workspace.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+def test_path_in_workspace():
+    # test relative path
+    relative_path = "data/input.txt"
+    expected_path = WORKSPACE_PATH / "data/input.txt"
+    assert path_in_workspace(relative_path) == expected_path
+
+    # test absolute path
+    absolute_path = Path("/home/user/data/input.txt")
+    expected_path = WORKSPACE_PATH / "home/user/data/input.txt"
+    assert path_in_workspace(absolute_path) == expected_path
```
Comment (Pwuts) [inline tests/test_workspace.py]: This isn't going to work. Joining anchored paths results in just the absolute path, so `path_in_workspace(Path('/home/user/data/input.txt'))` gives `Path('/home/user/data/input.txt')`

**Review by Pwuts — CHANGES_REQUESTED** [review]
Thanks for writing these tests!

**Comment by jacobyoby** [comment]
fixed (i think)

**Comment by Pwuts** [comment]
> someone committed a better version already

I just found out what you mean

**Comment by Pwuts** [comment]
Closing as the version of `workspace.py` on which this PR is based was deleted. Sorry. :(

Thanks anyways for submitting!

**Pull request closed without merging**